### PR TITLE
Remove trainer args from conf/defaults.yaml

### LINKED ISSUE #347: detect_anomaly not supported for pytorch-lightning<1.5
Support for passing `detect_anomaly` (in `conf/defaults.yaml`) was added in pytorch-lightning 1.5 (Refer PyTorchLightning/pytorch-lightning#9848). For lower versions training will fail with 

```
Traceback (most recent call last):
  File "/home/ashwin/Desktop/Projects/torchgeo/train.py", line 226, in <module>
    main(conf)
  File "/home/ashwin/Desktop/Projects/torchgeo/train.py", line 196, in main
    trainer = pl.Trainer(**trainer_args)
  File "/home/ashwin/anaconda3/envs/torchgeo/lib/python3.9/site-packages/pytorch_lightning/trainer/connectors/env_vars_connector.py", line 40, in insert_env_defaults
    return fn(self, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'detect_anomaly'
```
So we either need to pin minimum version to 1.5 or remove this flag from `conf/defaults.yaml`.

### PR DESCRIPTION
**Pull request by ashnair1**
Closes #347

### CODE REVIEW

**Comment by ashnair1** [comment]
1. While `defaults.yaml` pinned torchgeo to use newer versions of `pytorch-lightning` it was useful in documenting what values could be set. Now that it's being removed, users don't really have a reference to set up their training configs. For training args I guess they can refer `https://pytorch-lightning.readthedocs.io/en/<version>/common/trainer.html#init` (replace version with whatever version of pytorch-lightning they're using).

2. The experiment level arguments (seed, output_dir, log_dir) must be set in all task yamls for training to run. I can add the something like this to all the task_defaults configs.

```
program:  # These are experiment level arguments
  seed: 0
  experiment_name: blah
  overwrite: True
  output_dir: output
  data_dir: data
  log_dir: logs
```

File: train.py [hunk 1]
```diff
@@ -86,20 +86,13 @@ def set_up_omegaconf() -> DictConfig:
         FileNotFoundError: when ``config_file`` does not exist
         ValueError: when ``task.name`` is not a valid task
     """
```
Comment (adamjstewart) [inline train.py:88]: This docstring has a lot of references to `defaults.yaml` too

**Comment by adamjstewart** [comment]
2. Hmm, we may want to keep a `defaults.yaml` that only contains these defaults. @calebrob6 thoughts?

**Comment by ashnair1** [comment]
Keeping a `defaults.yaml` with just these defaults seem like a good idea since they're not related to any library. The issue in #347  was only with trainer args which was dependent on the version of PL.

**Comment by adamjstewart** [comment]
Some of those do actually make sense to override on a dataset-level. Like `experiment_name` and `*_dir` could be dataset-specific so all output doesn't end up in the same directory.

**Comment by adamjstewart** [comment]
Another option is to make all settings (not just the ones in `defaults.yaml`) optional using something like `getattr`.

**Comment by ashnair1** [comment]
> Some of those do actually make sense to override on a dataset-level. Like `experiment_name` and `*_dir` could be dataset-specific so all output doesn't end up in the same directory.


Agreed. Basically I was thinking to leave the first half of `conf/defaults.yaml` more or less as is. So this is all it would contain:

```
config_file: null  # This lets the user pass a config filename to load other arguments from

program:  # These are the arguments that define how the train.py script works
  seed: 0
  output_dir: output
  log_dir: logs
  overwrite: False
```
We only need to specify the **seed**, **output_dir** and **log_dir** because they're explicitly used in train.py. experiment_name would go in the specific task.yaml. I made a mistake including it before.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**